apple/Sources/FerrostarCore/Extensions/CoreLocation Extensions.swift

```diff
@@ -0,0 +1,147 @@
+    ///
+    /// - Parameters:
+    ///   - course: The direction the device is travelling, measured in degrees relative to true north.
+    ///   - courseAccuracy: The accuracy of the direction
+    public init?(course: CLLocationDirection,
+                 courseAccuracy: CLLocationDirectionAccuracy) {
+        guard course >= 0 && courseAccuracy >= 0 else {
+            return nil
+        }
+        
+        self.init(degrees: UInt16(course), accuracy: UInt16(courseAccuracy))
+    }
+}
+
+extension Heading {
+    
+    /// Initialize a Heading if it can be represented as integers
+    ///
+    /// This returns nil if the heading or accuracy  are invalid as defined by Apple (negative)
+    ///
+    /// - Parameter clHeading: The CoreLocation heading provided by the location manager.
+    public init?(clHeading: CLHeading) {
+        guard clHeading.trueHeading >= 0
+            && clHeading.headingAccuracy >= 0 else {
+            return nil
+        }
+        
+        self.init(trueHeading: UInt16(clHeading.trueHeading),
+                  accuracy: UInt16(clHeading.headingAccuracy),
+                  timestamp: clHeading.timestamp)
+    }
+}
+
+extension UserLocation {
+    
+    /// Initialize a UserLocation from values.
+    ///
+    /// - Parameters:
+    ///   - latitude: The latitude in decimal degrees
+    ///   - longitude: The longitude in decimal degrees
+    ///   - horizontalAccuracy: The horizontal accuracy
+    ///   - course: The direction of travel measured in degrees clockwise from north.
+    ///   - courseAccuracy: The course accuracy measured in degrees.
+    ///   - timestamp: The timestamp of the location record.
+    public init(latitude: CLLocationDegrees,
+                longitude: CLLocationDegrees,
+                horizontalAccuracy: CLLocationDistance,
+                course: CLLocationDirection,
+                courseAccuracy: CLLocationDirectionAccuracy,
+                timestamp: Date) {
+        
+        self.init(coordinates: GeographicCoordinate(lat: latitude, lng: longitude),
+                  horizontalAccuracy: horizontalAccuracy,
+                  courseOverGround: CourseOverGround(course: course, courseAccuracy: courseAccuracy),
+                  timestamp: timestamp)
+    }
+    
+    /// Initialize a UserLocation with a coordinate only.
+    ///
+    /// - Parameter clCoordinateLocation2D: A core location coordinate.
+    public init(clCoordinateLocation2D: CLLocationCoordinate2D) {
+        // This behavior matches how CLLocation initializes with a coordinate (setting accuracy to 0 & date to now)
+        self.init(coordinates: GeographicCoordinate(cl: clCoordinateLocation2D),
+                  horizontalAccuracy: 0,
+                  courseOverGround: nil,
+                  timestamp: Date())
+    }
+    
+    /// Initialize a UserLocation from an Apple CoreLocation CLLocation
+    ///
+    /// Unlike CourseOverGround & Heading, this value will init with invalid values.
```

Comment:
```suggestion
    /// Unlike CourseOverGround & Heading, this initializer will accept CLLocations that have invalid values.
```

Something like this... was a bit confusing on a first read ;)